Player movement controller

SwiffterAnim opened the pull request:
As mentioned, it's kinda working, but it breaks when they're mixed.

SwiffterAnim committed:
First Try on Movement

Jump is working, but haven't managed to get the aim working with the mouse position with the new input system yet.

SwiffterAnim committed:
Movement 'kinda' working

works well if using one of the other from the start. But not if you switch between both after the game has started.

omidizadi commented on Assets/Modules/Movement/Scripts/PlayerMovementController.cs:
```diff
@@ -0,0 +1,70 @@
+
+        // Checking if Gamepad is connected. Only does it at Start, so if gamepad is off, start play and THEN turn on gamepad, it will jump towards mouse.
+        // If starts with gamepad, then jumps with mouse, it explodes xD
+        foreach(var device in InputSystem.devices){
```
You should be able to listen to the device change in the input system and then run this loop whenever a new device is connected/disconnected:
Maybe in awake or start subscribe to the device change event:
```
 InputSystem.onDeviceChange += OnDeviceChange;
```
and then
```
private void OnDeviceChange(InputDevice device, InputDeviceChange change)
    {
        // Update input devices when a device is added or removed
        if (change == InputDeviceChange.Added || change == InputDeviceChange.Removed)
        {
            // run the loop
        }
    }
```

SwiffterAnim replied:
What does "subscribing to the device change event" mean? Is it like saying to the program "I want to hear from you in case the device list changes"?

Then I create that method "OnDeviceChange" that will make sure that only when 1 device is added or removed I run the loop, correct?
Thanks!

SwiffterAnim replied:
Also do I need the "InputDevice device" as a parameter of this OnDeviceChange method? because inside the "foreach" loop I'm already declaring that. I'm actually getting an error:
![Screenshot 2024-06-19 at 09 41 17](https://github.com/SwiffterAnim/Bounce-Escape/assets/170499803/5d6932bb-81b1-4f14-b553-8021ffd24758)

SwiffterAnim replied:
Sorry, another image for clarification:
![Screenshot 2024-06-19 at 09 42 55](https://github.com/SwiffterAnim/Bounce-Escape/assets/170499803/c3ee45cc-b18d-49ac-b2f8-3ca4c0da431a)

SwiffterAnim replied:
Deleting the InputDevice device from the parenthesis of the method, or changing the "var device" fixes the issue inside the method but breaks the 
        InputSystem.onDeviceChange += OnDeviceChange;

deleting the "var" in front of the "device" breaks the "in". it says that:
Type and identifier are both required in a foreach statement[CS0230]

the only thing that doesn't give me any errors is changing the name of that variable:

private void OnDeviceChange(InputDevice device, InputDeviceChange change)
    {
        // Update input devices when a device is added or removed
        if (change == InputDeviceChange.Added || change == InputDeviceChange.Removed)
        {
            foreach(var devices in InputSystem.devices){
            if(devices is Gamepad){
            gamepadIsConnected = true;
            break;
            }
        }
        }
    }
    
    
    ((Sorry I don't know how to put code nicely here...))

omidizadi replied:
> What does "subscribing to the device change event" mean? Is it like saying to the program "I want to hear from you in case the device list changes"?
> 
> Then I create that method "OnDeviceChange" that will make sure that only when 1 device is added or removed I run the loop, correct? Thanks!

You got the subscribing concept right. In this line
```
InputSystem.onDeviceChange += OnDeviceChange;
```
the += is like "please fire this function whenever a device changed.

omidizadi replied:
> Deleting the InputDevice device from the parenthesis of the method, or changing the "var device" fixes the issue inside the method but breaks the InputSystem.onDeviceChange += OnDeviceChange;
> 
> deleting the "var" in front of the "device" breaks the "in". it says that: Type and identifier are both required in a foreach statement[CS0230]
> 
> the only thing that doesn't give me any errors is changing the name of that variable:
> 
> private void OnDeviceChange(InputDevice device, InputDeviceChange change) { // Update input devices when a device is added or removed if (change == InputDeviceChange.Added || change == InputDeviceChange.Removed) { foreach(var devices in InputSystem.devices){ if(devices is Gamepad){ gamepadIsConnected = true; break; } } } }
> 
> ```
> ((Sorry I don't know how to put code nicely here...))
> ```

I see! The problem is that we have 2 variables with the same name `device`
hmm now that I look closer at this, you don't need to run the loop inside the onDeviceChange method. just check 
```
if(device is Joystick)
```
and use the device in the method's parameter. it should work

omidizadi replied:
to make it more pretty you can even do this one line only:
```
gamepadIsConnected = device is Joystick;
```

SwiffterAnim replied:
That's interesting about the subscribing! I didn't know about that concept! pretty cool!
And cool, I fixed the note.

There are a couple things in the movement that I'd like to make sure we do, but I'll let you decide the order of things ok?
1- I want him to not be able to double or triple jump. On my original, once he jumps, he can only jump again if he hits the wall, or if he connects into a hook, or if he gets hit by something.
2- I want the force/velocity to reset every time we jump, right now, if we jump to one side, then jump to the other side, it's like the physics of the rigidbody are adding or subtracting force from the already moving object. Meaning, if we jump to the left and then to the right in quick succession, right now the end result would be that he ends up still on the screen, rather than jumping to the right. I think I fixed this issue before by resetting the physics somehow on my prototype.

I'll also adjust the force of the jump and the gravity later on.

omidizadi replied:
> That's interesting about the subscribing! I didn't know about that concept! pretty cool! And cool, I fixed the note.
> 
> There are a couple things in the movement that I'd like to make sure we do, but I'll let you decide the order of things ok? 1- I want him to not be able to double or triple jump. On my original, once he jumps, he can only jump again if he hits the wall, or if he connects into a hook, or if he gets hit by something. 2- I want the force/velocity to reset every time we jump, right now, if we jump to one side, then jump to the other side, it's like the physics of the rigidbody are adding or subtracting force from the already moving object. Meaning, if we jump to the left and then to the right in quick succession, right now the end result would be that he ends up still on the screen, rather than jumping to the right. I think I fixed this issue before by resetting the physics somehow on my prototype.
> 
> I'll also adjust the force of the jump and the gravity later on.

Regarding point 1, you can easily do that using colliders and checking the collision in code (I guess that's how you did it last time) also you need to have a flag that shows if the player has jumped and locks the jump action until it hits the wall of a hook.
As for point 2, I guess you can also set the velocity to zero as soon as the collision is detected.
Since we still don't have the Obstacles implemented (hook/wall) just use the famous gameObject.CompareTag("") and check the tag "Obstacle" for now to identify collisions. If you don't know which type of collision you should check  and also how to setup the collider components, let me know 👍

SwiffterAnim replied:
Ok, I fixed both these issues!
One other issue that I'm having now, that I don't remember having on my prototype, is why doesn't the character bounce OFF the walls? I'm going against the wall and sticking to it, rather than bouncing off of it.. I would like to bounce off of it, like a pool ball against the walls of the pool table =')
I'm also having difficulties with the Mouse VS Gamepad thing, it's bugging completely. So I commented out on my latest version the possibility of aiming with the mouse and it's working better. We can try figuring this out later. I'm losing too much time with that =')

omidizadi replied:
hmm I think you need to create a `PhysicsMaterial` and assign it to the wall's Rigidbody. Then you can adjust the bounciness on the material. If it didn't work try to assign it to the player rigidbody as well.

SwiffterAnim replied:
Yes that's what it was! Thanks hehe

SwiffterAnim committed:
Updating code with OnDeviceChange method

SwiffterAnim committed:
Quick Update on OnDeviceChange method.

SwiffterAnim committed:
Jump fixes

Added canJump so the player can only jump once or after it collides with something.
Also fixed the rigidbody issue and put the velocity to zero before jumping again.

omidizadi commented on Assets/Modules/Movement/Scripts/PlayerMovementController.cs:
```diff
@@ -68,6 +79,17 @@ private void OnJump(InputValue value)
+
+    private void OnCollisionEnter2D(Collision2D other)
+    {
+        canJump = true;
```
Does this work without setting the `rb.velocity` to zero?

SwiffterAnim replied:
the jump works, but it's that thing I mentioned before, that if I don't set it up to zero before jumping again and try to jump the opposite direction, rather than full jump to the other side, the forces are sort of cancelling each other out rather than jumping? putting it to zero fixes that issue.

omidizadi commented on Assets/Modules/Movement/Scripts/PlayerMovementController.cs:
```diff
@@ -0,0 +1,95 @@
+        // Update input devices when a device is added or removed
+        if (change == InputDeviceChange.Added || change == InputDeviceChange.Removed)
+        {
+            gamepadIsConnected = device is Gamepad; //same as if(device is Gamepad){gamepadIsConnected = true;}
```
I'm not 100% sure but I think here you should check if the device is `Joystick` If you wanna check the controller 🤔 try it and see if it solves the input issue. If it didn't work I'll check it myself.

SwiffterAnim replied:
Changing it to joystick does not solve the issue for me, the issue is that the mouse is still affecting the jump, even if I'm using the controller (even if I'm constantly moving the stick of the controller, which assumes to me the value of the stick is always changing, thus, the program should always detect it) but from time to time it still follows the mouse.
So Gamepad or Joystick both do the same and have the same issue unfortunately!

omidizadi commented on Assets/Modules/Obstacles/Scripts/ObstacleEntity.cs:
```diff
@@ -0,0 +1,20 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class ObstacleEntity : MonoBehaviour
```
This is a good start. The toggles are indeed Serializefields that we can switch on and off in the inspector

omidizadi replied:
Also I think something went wrong when you wanted to create a new branch for the obstacle entity implementation and it was pushed here. No problem at all. We can merge this branch and move forward with the new one :)

SwiffterAnim replied:
Ah yes, because I was going to work on the Obstacle entity, but then decided to go back to the movement, and I wasn't really sure how to do that haha so I just changed branches and brought the changes to this branch. sorry!

omidizadi commented on Assets/Modules/Obstacles/Scripts/ObstacleEntity.cs:
```diff
@@ -0,0 +1,20 @@
+
+public class ObstacleEntity : MonoBehaviour
+{
+    [SerializeField] bool hooknessActivated;
```
don't forget `private` keywords for variables and methods

SwiffterAnim committed:
Some tests before moving on.

Tried Joystick rather than gamepad, issue is the same.
But added bounciness to the wall which is good.
